Refactor/settings export

Leptopoda opened the pull request:
Closes #496
Closes #173 

Compleatly re implements the `SettingsExportHelper`.

---

On a sidenote: Why are we base64 encoding the exported data? wouldn't it be sufficient enough to safe the json directly?

Leptopoda commented:
Ci will be pasing with https://github.com/provokateurin/nextcloud-neon/pull/503

Leptopoda commented:
I somehow dislike this :(
I think we should add an `Exportable` interface or something like that and let the AccountsBloc or whatever else we'll have define what they want to export. The downside is that the exporting/importing logic wouldn't be in one place anymore.

Leptopoda commented:
I somehow dislike this :(
I think we should add an `Exportable` interface or something like that and let the AccountsBloc or whatever else we'll have define what they want to export. The downside is that the exporting/importing logic wouldn't be in one place anymore.

provokateurin commented:
The base64 encoding makes no sense, no idea why it is there. I like the exportable interface :+1:

provokateurin commented on packages/neon/neon/lib/src/settings/models/exportable.dart:
```diff
@@ -0,0 +1,35 @@
+
+/// Exportable data interface.
+abstract interface class Exportable {
+  /// Exports this into a json map entry.
```
```suggestion
  /// Exports into a json map entry.
```

Leptopoda replied:
Why without the this?

provokateurin commented on packages/neon/neon/lib/src/settings/models/exportable.dart:
```diff
@@ -0,0 +1,35 @@
+
+/// Serialization helpers for a collection of [Option]s.
+extension SerializeOptions on Iterable<Option<dynamic>> {
+  /// Serializes this into  an [Iterable<JsonEntry>].
```
```suggestion
  /// Serializes into an [Iterable<JsonEntry>].
```

provokateurin commented on packages/neon/neon/lib/src/settings/models/nextcloud_app_options.dart:
```diff
@@ -1,23 +1,66 @@
-  NextcloudAppOptions(this.storage);
+/// Collection of [Option]s.
+abstract class OptionsCollection implements Exportable {
+  /// Creates a new OptionsCollection.
```
Not really helpful, I usually just ignore constructors

Leptopoda replied:
Should I remove them?

provokateurin replied:
Yes please

provokateurin commented on packages/neon/neon/lib/src/settings/models/nextcloud_app_options.dart:
```diff
@@ -1,23 +1,66 @@
+  }
+}
+
+/// OpptionsCollection for a neon app.
```
```suggestion
/// OpptionsCollection for an [AppImplementation].
```

Is this better?

Leptopoda replied:
The thing is that `AppImplementation` is not visible in that scope and just importing it for the reference in the comment doesn't feel right

provokateurin replied:
Alright

provokateurin commented on packages/neon/neon/lib/src/settings/models/nextcloud_app_options.dart:
```diff
@@ -1,23 +1,66 @@
+
+/// OpptionsCollection for a neon app.
+abstract class NextcloudAppOptions extends OptionsCollection {
+  /// Creates a new NextcloudAppOptions.
```
Also not helpful imo

provokateurin commented on packages/neon/neon/lib/src/settings/models/option.dart:
```diff
@@ -83,6 +83,15 @@ abstract class Option<T> extends ChangeNotifier implements ValueListenable<T> {
     value = defaultValue;
   }
 
+  /// loads [data] into [value] by calling [deserialize] on it.
```
```suggestion
  /// Loads [data] into [value] by calling [deserialize] on it.
```

provokateurin reviewed (changes requested):
Importing settings does not work correctly for apps and accounts. The export definitely includes the correct values.

Leptopoda commented:
I also took the chance and cleaned up our AppStorage :)

provokateurin commented:
Now apps and accounts work, but the global options like theme all don't work :grimacing:

provokateurin commented:
I think unit tests for this would be great otherwise it will probably break very quickly again :/

Leptopoda commented:
Unit tests are hard especially around our blocs :(

Leptopoda commented:
My bad The initial version was hard to unit test but the current code should make it fairly simple.

I'll add a few test later.

Leptopoda commented:
@provokateurin  I've added some unit tests and also updated the `Option` tests to get 100% coverage again.

I didn't test the specifics of the options though as they are to entangled with the blocs for now.

provokateurin commented:
Please rebase then I can properly review :)

provokateurin reviewed (changes requested):
Importing account settings now works, but app settings are still broken.

Leptopoda commented:
Ok I just rebased onto main (aka #688 as main would be broken) and I can't reproduce the issue.
I tested it with every app, GlobalOptions and AccountOptions (It would be very interesting as AFICT the export part has 100% coverage and I'm fairly sure the tests are not false).

Can you please verify that you had the latest version of the PR checked out?
Can you also provide further information of what exact settings are broken or even provide the exported json.

Here is the export I successfully verified as working:
[nextcloud-neon-settings-1693437863.json.txt](https://github.com/nextcloud/neon/files/12481254/nextcloud-neon-settings-1693437863.json.txt)

provokateurin reviewed (approved):
You are right, it was working correctly. I just so happened to find https://github.com/nextcloud/neon/pull/689 by accident and thought the import wasn't working for that option. Somewhat lucky that I found it.
